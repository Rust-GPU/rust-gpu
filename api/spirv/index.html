<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The SPIR-V header."><title>spirv - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv/lib.rs.html#1-14">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The SPIR-V header.</p>
<p>This crate contains Rust definitions of all SPIR-V structs, enums,
and constants.</p>
<p>The version of this crate is the version of SPIR-V it contains.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.CooperativeMatrixOperands.html" title="struct spirv::CooperativeMatrixOperands">CooperativeMatrixOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_operands_a_cooperative_matrix_operands">CooperativeMatrixOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.FPFastMathMode.html" title="struct spirv::FPFastMathMode">FPFastMathMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode">FPFastMathMode</a></div></li><li><div class="item-name"><a class="struct" href="struct.FragmentShadingRate.html" title="struct spirv::FragmentShadingRate">FragmentShadingRate</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fragment_shading_rate_a_fragment_shading_rate">FragmentShadingRate</a></div></li><li><div class="item-name"><a class="struct" href="struct.FunctionControl.html" title="struct spirv::FunctionControl">FunctionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_control_a_function_control">FunctionControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.ImageOperands.html" title="struct spirv::ImageOperands">ImageOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_operands_a_image_operands">ImageOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.KernelProfilingInfo.html" title="struct spirv::KernelProfilingInfo">KernelProfilingInfo</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_profiling_info_a_kernel_profiling_info">KernelProfilingInfo</a></div></li><li><div class="item-name"><a class="struct" href="struct.LoopControl.html" title="struct spirv::LoopControl">LoopControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_loop_control_a_loop_control">LoopControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemoryAccess.html" title="struct spirv::MemoryAccess">MemoryAccess</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_access_a_memory_access">MemoryAccess</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemorySemantics.html" title="struct spirv::MemorySemantics">MemorySemantics</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics_a_memory_semantics">MemorySemantics</a></div></li><li><div class="item-name"><a class="struct" href="struct.RayFlags.html" title="struct spirv::RayFlags">RayFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_flags_a_ray_flags">RayFlags</a></div></li><li><div class="item-name"><a class="struct" href="struct.SelectionControl.html" title="struct spirv::SelectionControl">SelectionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_selection_control_a_selection_control">SelectionControl</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.AccessQualifier.html" title="enum spirv::AccessQualifier">AccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_access_qualifier_a_access_qualifier">AccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.AddressingModel.html" title="enum spirv::AddressingModel">AddressingModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model">AddressingModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.BuiltIn.html" title="enum spirv::BuiltIn">BuiltIn</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_built_in_a_built_in">BuiltIn</a></div></li><li><div class="item-name"><a class="enum" href="enum.CLOp.html" title="enum spirv::CLOp">CLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.ExtendedInstructionSet.100.html">OpenCL.std</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.Capability.html" title="enum spirv::Capability">Capability</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability">Capability</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixLayout.html" title="enum spirv::CooperativeMatrixLayout">CooperativeMatrixLayout</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_layout_a_cooperative_matrix_layout">CooperativeMatrixLayout</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixUse.html" title="enum spirv::CooperativeMatrixUse">CooperativeMatrixUse</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_use_a_cooperative_matrix_use">CooperativeMatrixUse</a></div></li><li><div class="item-name"><a class="enum" href="enum.Decoration.html" title="enum spirv::Decoration">Decoration</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_decoration_a_decoration">Decoration</a></div></li><li><div class="item-name"><a class="enum" href="enum.Dim.html" title="enum spirv::Dim">Dim</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_dim_a_dim">Dim</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionMode.html" title="enum spirv::ExecutionMode">ExecutionMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_mode_a_execution_mode">ExecutionMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionModel.html" title="enum spirv::ExecutionModel">ExecutionModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_model_a_execution_model">ExecutionModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPDenormMode.html" title="enum spirv::FPDenormMode">FPDenormMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_denorm_mode_a_fp_denorm_mode">FPDenormMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPOperationMode.html" title="enum spirv::FPOperationMode">FPOperationMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_operation_mode_a_fp_operation_mode">FPOperationMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPRoundingMode.html" title="enum spirv::FPRoundingMode">FPRoundingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_rounding_mode_a_fp_rounding_mode">FPRoundingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FunctionParameterAttribute.html" title="enum spirv::FunctionParameterAttribute">FunctionParameterAttribute</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_parameter_attribute_a_function_parameter_attribute">FunctionParameterAttribute</a></div></li><li><div class="item-name"><a class="enum" href="enum.GLOp.html" title="enum spirv::GLOp">GLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/GLSL.std.450.html">GLSL.std.450</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.GroupOperation.html" title="enum spirv::GroupOperation">GroupOperation</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_group_operation_a_group_operation">GroupOperation</a></div></li><li><div class="item-name"><a class="enum" href="enum.HostAccessQualifier.html" title="enum spirv::HostAccessQualifier">HostAccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_host_access_qualifier_a_host_access_qualifier">HostAccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelDataType.html" title="enum spirv::ImageChannelDataType">ImageChannelDataType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_data_type_a_image_channel_data_type">ImageChannelDataType</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelOrder.html" title="enum spirv::ImageChannelOrder">ImageChannelOrder</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_order_a_image_channel_order">ImageChannelOrder</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageFormat.html" title="enum spirv::ImageFormat">ImageFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format">ImageFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.InitializationModeQualifier.html" title="enum spirv::InitializationModeQualifier">InitializationModeQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_initialization_mode_qualifier_a_initialization_mode_qualifier">InitializationModeQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.KernelEnqueueFlags.html" title="enum spirv::KernelEnqueueFlags">KernelEnqueueFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_enqueue_flags_a_kernel_enqueue_flags">KernelEnqueueFlags</a></div></li><li><div class="item-name"><a class="enum" href="enum.LinkageType.html" title="enum spirv::LinkageType">LinkageType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_linkage_type_a_linkage_type">LinkageType</a></div></li><li><div class="item-name"><a class="enum" href="enum.LoadCacheControl.html" title="enum spirv::LoadCacheControl">LoadCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_load_cache_control_a_load_cache_control">LoadCacheControl</a></div></li><li><div class="item-name"><a class="enum" href="enum.MemoryModel.html" title="enum spirv::MemoryModel">MemoryModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_model_a_memory_model">MemoryModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.Op.html" title="enum spirv::Op">Op</a></div><div class="desc docblock-short">SPIR-V <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_instructions_a_instructions">instructions</a> opcodes</div></li><li><div class="item-name"><a class="enum" href="enum.OverflowModes.html" title="enum spirv::OverflowModes">OverflowModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_overflow_modes_a_overflow_modes">OverflowModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.PackedVectorFormat.html" title="enum spirv::PackedVectorFormat">PackedVectorFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_packed_vector_format_a_packed_vector_format">PackedVectorFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.QuantizationModes.html" title="enum spirv::QuantizationModes">QuantizationModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_quantization_modes_a_quantization_modes">QuantizationModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCandidateIntersectionType.html" title="enum spirv::RayQueryCandidateIntersectionType">RayQueryCandidateIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_candidate_intersection_type_a_ray_query_candidate_intersection_type">RayQueryCandidateIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCommittedIntersectionType.html" title="enum spirv::RayQueryCommittedIntersectionType">RayQueryCommittedIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_committed_intersection_type_a_ray_query_committed_intersection_type">RayQueryCommittedIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryIntersection.html" title="enum spirv::RayQueryIntersection">RayQueryIntersection</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_intersection_a_ray_query_intersection">RayQueryIntersection</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerAddressingMode.html" title="enum spirv::SamplerAddressingMode">SamplerAddressingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_addressing_mode_a_sampler_addressing_mode">SamplerAddressingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerFilterMode.html" title="enum spirv::SamplerFilterMode">SamplerFilterMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_filter_mode_a_sampler_filter_mode">SamplerFilterMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.Scope.html" title="enum spirv::Scope">Scope</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_scope_a_scope">Scope</a></div></li><li><div class="item-name"><a class="enum" href="enum.SourceLanguage.html" title="enum spirv::SourceLanguage">SourceLanguage</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_source_language_a_source_language">SourceLanguage</a></div></li><li><div class="item-name"><a class="enum" href="enum.StorageClass.html" title="enum spirv::StorageClass">StorageClass</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_storage_class_a_storage_class">StorageClass</a></div></li><li><div class="item-name"><a class="enum" href="enum.StoreCacheControl.html" title="enum spirv::StoreCacheControl">StoreCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_store_cache_control_a_store_cache_control">StoreCacheControl</a></div></li></ul><h2 id="constants" class="section-header">Constants<a href="#constants" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="constant" href="constant.MAGIC_NUMBER.html" title="constant spirv::MAGIC_NUMBER">MAGIC_NUMBER</a></div></li><li><div class="item-name"><a class="constant" href="constant.MAJOR_VERSION.html" title="constant spirv::MAJOR_VERSION">MAJOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.MINOR_VERSION.html" title="constant spirv::MINOR_VERSION">MINOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.REVISION.html" title="constant spirv::REVISION">REVISION</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.Word.html" title="type spirv::Word">Word</a></div></li></ul></section></div></main></body></html>