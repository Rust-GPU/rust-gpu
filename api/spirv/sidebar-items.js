window.SIDEBAR_ITEMS = {"constant":["MAGIC_NUMBER","MAJOR_VERSION","MINOR_VERSION","REVISION"],"enum":["AccessQualifier","AddressingModel","BuiltIn","CLOp","Capability","CooperativeMatrixLayout","CooperativeMatrixUse","Decoration","Dim","ExecutionMode","ExecutionModel","FPDenormMode","FPOperationMode","FPRoundingMode","FunctionParameterAttribute","GLOp","GroupOperation","HostAccessQualifier","ImageChannelDataType","ImageChannelOrder","ImageFormat","InitializationModeQualifier","KernelEnqueueFlags","LinkageType","LoadCacheControl","MemoryModel","Op","OverflowModes","PackedVectorFormat","QuantizationModes","RayQueryCandidateIntersectionType","RayQueryCommittedIntersectionType","RayQueryIntersection","SamplerAddressingMode","SamplerFilterMode","Scope","SourceLanguage","StorageClass","StoreCacheControl"],"struct":["CooperativeMatrixOperands","FPFastMathMode","FragmentShadingRate","FunctionControl","ImageOperands","KernelProfilingInfo","LoopControl","MemoryAccess","MemorySemantics","RayFlags","SelectionControl"],"type":["Word"]};