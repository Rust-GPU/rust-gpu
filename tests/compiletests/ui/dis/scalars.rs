#![crate_name = "scalars"]

// Tests all the (supported) Rust integer/floating-point scalar types.

// revisions: supported nocaps
//[supported] build-pass
//[supported] compile-flags: -C target-feature=+Int8,+Int16,+Int64,+Float64
//[nocaps] build-fail

// compile-flags: -C llvm-args=--disassemble-globals
// normalize-stderr-test "OpCapability VulkanMemoryModel\n" -> ""
// normalize-stderr-test "OpSource .*\n" -> ""
// normalize-stderr-test "OpExtension .SPV_KHR_vulkan_memory_model.\n" -> ""
// normalize-stderr-test "OpMemoryModel Logical Vulkan" -> "OpMemoryModel Logical Simple"

// HACK(eddyb) `compiletest` handles `ui\dis\`, but not `ui\\dis\\`, on Windows.
// normalize-stderr-test "ui/dis/" -> "$$DIR/"

use spirv_std::spirv;

#[spirv(fragment)]
pub fn main(
    out: &mut u32,

    #[spirv(flat)] in_u8: u8,
    #[spirv(flat)] in_u16: u16,
    #[spirv(flat)] in_u32: u32,
    #[spirv(flat)] in_u64: u64,
    // FIXME(eddyb) support `u128` somehow!
    #[cfg(not(supported))]
    #[spirv(flat)]
    in_u128: u128,

    #[spirv(flat)] in_i8: i8,
    #[spirv(flat)] in_i16: i16,
    #[spirv(flat)] in_i32: i32,
    #[spirv(flat)] in_i64: i64,
    // FIXME(eddyb) support `i128` somehow!
    #[cfg(not(supported))]
    #[spirv(flat)]
    in_i128: i128,

    in_f32: f32,
    #[spirv(flat)] in_f64: f64,
) {
    // HACK(eddyb) to make it more obvious in the disassembly, each type gets a
    // constant with its bit width, disambiguated for signed integers by negation.
    *out |= (in_u8 * 8) as u32;
    *out |= (in_u16 * 16) as u32;
    *out |= (in_u32 * 32) as u32;
    *out |= (in_u64 * 64) as u32;
    // FIXME(eddyb) support `u128` somehow!
    #[cfg(not(supported))]
    {
        // FIXME(eddyb) constants still emit zombies that get reported too early.
        // *out |= (in_u128 * 128) as u32;
        *out |= (in_u128 + in_u128) as u32;
    }

    *out |= (in_i8 * -8) as u32;
    *out |= (in_i16 * -16) as u32;
    *out |= (in_i32 * -32) as u32;
    *out |= (in_i64 * -64) as u32;
    // FIXME(eddyb) support `i128` somehow!
    #[cfg(not(supported))]
    {
        // FIXME(eddyb) constants still emit zombies that get reported too early.
        // *out |= (in_i128 * -128) as u32;
        *out |= (in_i128 + in_i128) as u32;
    }

    *out |= (in_f32 * 32.0) as u32;
    *out |= (in_f64 * 64.0) as u32;
}
