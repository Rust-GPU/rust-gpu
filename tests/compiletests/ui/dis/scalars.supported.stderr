OpCapability Shader
OpCapability Float64
OpCapability Int64
OpCapability Int16
OpCapability Int8
OpMemoryModel Logical Simple
OpEntryPoint Fragment %1 "main" %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12
OpExecutionMode %1 OriginUpperLeft
%13 = OpString "$DIR/scalars.rs"
OpName %3 "in_u8"
OpName %4 "in_u16"
OpName %5 "in_u32"
OpName %6 "in_u64"
OpName %7 "in_i8"
OpName %8 "in_i16"
OpName %9 "in_i32"
OpName %10 "in_i64"
OpName %11 "in_f32"
OpName %12 "in_f64"
OpName %2 "out"
OpDecorate %3 Flat
OpDecorate %3 Location 0
OpDecorate %4 Flat
OpDecorate %4 Location 1
OpDecorate %5 Flat
OpDecorate %5 Location 2
OpDecorate %6 Flat
OpDecorate %6 Location 3
OpDecorate %7 Flat
OpDecorate %7 Location 4
OpDecorate %8 Flat
OpDecorate %8 Location 5
OpDecorate %9 Flat
OpDecorate %9 Location 6
OpDecorate %10 Flat
OpDecorate %10 Location 7
OpDecorate %11 Location 8
OpDecorate %12 Flat
OpDecorate %12 Location 9
OpDecorate %2 Location 0
%14 = OpTypeInt 32 0
%15 = OpTypePointer Output %14
%16 = OpTypeInt 8 0
%17 = OpTypePointer Input %16
%18 = OpTypeInt 16 0
%19 = OpTypePointer Input %18
%20 = OpTypePointer Input %14
%21 = OpTypeInt 64 0
%22 = OpTypePointer Input %21
%23 = OpTypeInt 8 1
%24 = OpTypePointer Input %23
%25 = OpTypeInt 16 1
%26 = OpTypePointer Input %25
%27 = OpTypeInt 32 1
%28 = OpTypePointer Input %27
%29 = OpTypeInt 64 1
%30 = OpTypePointer Input %29
%31 = OpTypeFloat 32
%32 = OpTypePointer Input %31
%33 = OpTypeFloat 64
%34 = OpTypePointer Input %33
%35 = OpTypeVoid
%36 = OpTypeFunction %35
%3 = OpVariable  %17  Input
%4 = OpVariable  %19  Input
%5 = OpVariable  %20  Input
%6 = OpVariable  %22  Input
%7 = OpVariable  %24  Input
%8 = OpVariable  %26  Input
%9 = OpVariable  %28  Input
%10 = OpVariable  %30  Input
%11 = OpVariable  %32  Input
%12 = OpVariable  %34  Input
%37 = OpConstant  %16  8
%2 = OpVariable  %15  Output
%38 = OpConstant  %18  16
%39 = OpConstant  %14  32
%40 = OpConstant  %21  64
%41 = OpConstant  %23  4294967288
%42 = OpConstant  %25  4294967280
%43 = OpConstant  %27  4294967264
%44 = OpConstant  %29  18446744073709551552
%45 = OpConstant  %31  1107296256
%46 = OpConstant  %14  0
%47 = OpConstant  %14  1333788671
%48 = OpTypeBool
%49 = OpConstant  %14  4294967295
%50 = OpConstant  %33  4634204016564240384
%51 = OpConstant  %21  0
%52 = OpConstant  %21  4751297606873776128
